docs/torchelastic-operator-design.md

```diff
@@ -0,0 +1,169 @@
+# TorchElastic Operator on Kubernetes
+
+## Background
+
+PyTorch continues to be used for the latest state-of-the-art research, making up nearly 70% of [papers](https://chillee.github.io/pytorch-vs-tensorflow/) that cite a framework.
+
+The current PyTorch Distributed Data Parallel (DDP) module enables data parallel training where each process trains the same model but on different shards of data. It enables bulk synchronous, multi-host, multi-GPU/CPU execution of ML training. However, DDP has several shortcomings; e.g.jobs cannot start without acquiring all the requested nodes; jobs cannot continue after a node fails due to an error or transient issue; jobs cannot incorporate a node that joined later; and lastly; progress cannot be made with the presence of a slow/stuck node.
+
+The focus of [PyTorch Elastic](https://github.com/pytorch/elastic), which uses Elastic Distributed Data Parallelism, is to address these issues and build a generic framework/APIs for PyTorch to enable reliable and elastic execution of these data parallel training workloads. It will provide better programmability, higher resilience to failures of all kinds, higher-efficiency and larger-scale training compared with pure DDP.
+
+## Motivation
+
+With job fault tolerance and elastic training, we can unlock a lot of features.
+
+Users can enable job priority and preemption in the cluster. Losing a task becomes acceptable and the user won't lose the entire job progress. More importantly, it will help guarantee SLAs of crititcal jobs, even in a cluster under resource pressure.
+
+Cost and GPU utilization will be further optimized with this feature, since users can launch jobs with partial resources and spot GPU instances can be used as well without worrying.
+
+## User Experience
+
+* Users should define the `min` and `max` number of tasks of a job instead of a fixed number. The TorchElastic Operator will launch jobs in Kubernetes, setup the needed network topology and manage the job lifecycle.
+* Users have the option to set the etcd endpoint used as the RDZV service for task coordination.
+* The desired `size`, being number of tasks, is an optional field in the range from `min` to `max`.
+* Users can easily create/delete a torch elastic job using `kubectl` using a job manifest.
+* Users are able to describe custom resources to monitor the job status.
+
+## High Level Design
+
+Torch elastic jobs does not have a concept of a parameter server, instead rank 0 (one of the workers) will become the leader. We can roughly think all the workers are equivalent and their communication is peer to peer.
```

Comment:
to be super pedantic, torchelastic doesn't strictly require a "leader", however most applications written on top of torchelastic will require some notion of leader. An example of a job that would not require a leader is a batch inference job (each worker will process `input_partition_k` and write the predictions out to `output_partition_k`).

Reply:
Thanks for correction. If this is transparent to user and we should not mention it.

Reply:
One more question, in the logs, besides rank  and communication to rdzvEndpoint,
I think rest training logs are identical in all pods? If we want to leave one pod for user to easily check logs, we can pick any of the workers?

Reply:
Ah, that would be application dependent. The imagenet example is written in a way that for everything else other than the `state.sync()` method, all ranks do exactly the same thing.

The example publishes logs to a CloudWatch where each worker gets its own log stream, so currently it is easy for the user to go check the logs for any worker. Currently, it is up to the user to set this up when writing their own jobs, but with some "deep-integration" we can imagine a world where when a user runs PyTorch Elastic on EKS/SageMaker, the logs/metrics by default go to CloudWatch, or at least to S3 with a CLI tool to pull the logs to local disk and inspect them. On our side, we could standardize on the log format so that it is parser friendly.

Reply:
Then we leave all pods when job is done for now. Manual job deletion will delete all pods belong to it. This allows log shipper to sink to centralized log solutions.